ibis/backends/dask/core.py

```diff
@@ -0,0 +1,144 @@
+different times during the loop.
+
+1. ``compute_time_context``
+---------------------------
+First, at the beginning of the main execution loop, ``compute_time_context`` is
+called. This function computes time contexts, and pass them to all children of
+the current node. These time contexts could be used in later steps to get data.
+This is essential for time series TableExpr, and related operations that adjust
+time context, such as window, asof_join, etc.
+
+By default, this function simply pass the unchanged time context to all
+children nodes.
+
+
+2. ``pre_execute``
+------------------
+Second, ``pre_execute`` is called.
+This function serves a similar purpose to ``data_preload``, the key difference
+being that ``pre_execute`` is called *every time* there's a call to execute.
+
+By default this function does nothing.
+
+3. ``execute_node``
+-------------------
+
+Then, when an expression is ready to be evaluated we call
+:func:`~ibis.dask.core.execute` on the expressions arguments and then
+:func:`~ibis.dask.dispatch.execute_node` on the expression with its
+now-materialized arguments.
+
+4. ``post_execute``
+-------------------
+The final step--``post_execute``--is called immediately after the previous call
+to ``execute_node`` and takes the instance of the
+:class:`~ibis.expr.operations.Node` just computed and the result of the
+computation.
+
+The purpose of this function is to allow additional computation to happen in
+the context of the current level of the execution loop. You might be wondering
+That may sound vague, so let's look at an example.
+
+Let's say you want to take a three day rolling average, and you want to include
+3 days of data prior to the first date of the input. You don't want to see that
+data in the result for a few reasons, one of which is that it would break the
+contract of window functions: given N rows of input there are N rows of output.
+
+Defining a ``post_execute`` rule for :class:`~ibis.expr.operations.WindowOp`
+allows you to encode such logic. One might want to implement this using
+:class:`~ibis.expr.operations.ScalarParameter`, in which case the ``scope``
+passed to ``post_execute`` would be the bound values passed in at the time the
+``execute`` method was called.
+
+
+Scope
+-----
+Scope is used across the execution phases, it iss a map that maps Ibis
+operators to actual data. It is used to cache data for calculated ops. It is
+an optimization to reused executed results.
+
+With time context included, the key is op associated with each expression;
+And scope value is another key-value map:
+- value: pd.DataFrame or pd.Series that is the result of executing key op
+- timecontext: of type TimeContext, the time context associated with the data
+stored in value
+
+See ibis.common.scope for details about the implementaion.
+"""
+
+from __future__ import absolute_import
+
+import dask.dataframe as dd
+
+from ibis.backends.pandas.core import (  # noqa: F401
```

Comment:
same here, let's try to remove the noqa and only import things we are actually using.